rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,199 @@
+we will create two apps: `testApp-tls-main-cluster-prod`, `testApp-postgres-main-cluster-prod`. Names for ports are mandatory in Kubernetes if more
+than one port is exposed for the service, so we always will be able to use it if we need it. User will be able to use annotation to specify
+preferred port, in that case only app with this port will be created, without port info in the app name.
+
+In order for discovery service to be able to dynamically create/update/delete apps we will add Read/Write permissions 
+for the apps to the built-in system role `Discovery`.
+
+Labels from the Kubernetes service will be copied to the corresponding Teleport app.
+
+## UX
+
+### Discovery
+
+Discovery service will use new `kubernetes` matchers to periodically poll list of services inside the Kubernetes cluster.
+Labels specified in the config will be used to filter out services that should be exposed. Also users will be able to
+specify namespaces to process. Another new field for the configuration of the discovery service will be `kubernetes_cluster` - 
+it specifies kubernetes cluster name in which discovery service is running. It is mandatory if `kubernetes` matchers are
+specified in the config.
+
+```yaml
+## This section configures the Discovery Service
+discovery_service:
+  enabled: yes
+  discovery_group: kube-auto-apps
+  kubernetes_cluster: main-cluster
+  kubernetes:
+    - types: ["app"] # in the future "db" will be possible
+      namespaces: [ "toronto", "porto" ]
+      labels:
+        env: staging
+    - types: ["app"]
+      namespaces: [ "seattle", "oakland" ]
+      labels:
+        env: testing
+```
+
+`kubernetes_cluster` field in the config will be translated into `teleport.dev/kubernetes-cluster` label on the Teleport app resource, so
+later it's be easier to target it in the corresponding app service.
+
+Kubernetes annotation `teleport.dev/discovery-protocol` controls protocol for the access of the Teleport app we create. If annotation is missing
+`tcp` type will be used in the app's URI. Additionally, if kubernetes service port definition has `appProtocol` field, and it contains
+values `http`/`https` we will use it in the URI, but annotation supersedes hint from the `appProtocol`.
+
+Annotation `teleport.dev/discovery-port` will control preferred port for the Kubernetes service. Its value should be one of the
+exposed service ports; otherwise, the app will not be imported. Value can be matched either by numeric value or by the name of
+the port defined on the service.
+
+Example Kubernetes service configs using these annotations can look like this:
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: first-service
+  labels:
+    app: first-service
+  annotations:
+    teleport.dev/discovery-type: app # Allowed values are [`app`]
+    teleport.dev/discovery-protocol: http # Allowed values are [`http`, `https`, `tcp`]
+    teleport.dev/discovery-port: 80
```

Comment:
How does the annotation approach work in case of multiple ports? According to the RFD:

> If Kubernetes service has multiple ports in use, we will expose each port as a separate app, with name of the app including the port's name specified in the service.

If I'm a user and have a service exposing non-standard ports for an HTTP and TCP app, I'll want to set the protocol for both.

Some known approaches are:
- putting a JSON in the annotation value (GKE does that for the CNLB stuff)
- putting a comma-separated list when we need to pass multiple items
- putting the port name in the annotation key

Reply:
Yes, I think we can make it comma separated list for the port and/or make `discovery-protocol` to be able to map, like `teleport.dev/discovery-protocol: "80:http,4242:https,4243:https"`. But I would like to keep first version simple and iterate on complicating annotations later. Also, as RFD mentions, `appProtocol` field on kubernetes port can be used to specify it per port in multiport setup.